numpy/_core/src/multiarray/mapping.c

```diff
@@ -1960,6 +1960,10 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
         tmp_arr = (PyArrayObject *)op;
     }
 
+    if (tmp_arr && solve_may_share_memory(self, tmp_arr, 1) == 1) {
```

Comment:
```suggestion
    if (tmp_arr && solve_may_share_memory(self, tmp_arr, 1) != 0) {
```
Since `< 0` means "failed to solve", which would mean we should do a defensive copy.

We will be missing if the index has overlap, such as something awkward like:
```
a = np.arange(10)
a[2:][a[:-2]] = 3
```
but let's just accept that here and make a new issue for that.